src/lib.rs

```diff
@@ -145,6 +156,47 @@ impl<'data> Iterator for DataFormatDescriptorIterator<'data> {
     }
 }
 
+struct KeyValueDataIterator<'data> {
+    data: &'data [u8],
+}
+
+impl<'data> Iterator for KeyValueDataIterator<'data> {
+    type Item = (&'data str, &'data [u8]);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut offset = 0;
+
+        let length = bytes_to_u32(self.data, &mut offset).ok()?;
+
+        let key_and_value = &self.data[offset..offset + length as usize];
+
+        offset += length as usize;
+
+        // Ensure that we're 4-byte aligned
+        if offset % 4 != 0 {
+            offset += 4 - (offset % 4);
+        }
+
+        self.data = &self.data[offset..];
+
+        // The key is terminated with a NUL character.
+        let key_end_index = match key_and_value.iter().position(|&c| c == b'\0') {
+            Some(key_and_value) => key_and_value,
+            None => return Some(("", &[])),
+        };
+
+        let key = &key_and_value[..key_end_index];
+        let value = &key_and_value[key_end_index + 1..];
+
+        let key = match std::str::from_utf8(key) {
+            Ok(key) => key,
+            Err(_) => return Some(("", value)),
```

Comment:
Another idea: Here and above, rather than returning weird data, how about we just skip the entry? That could be accomplished by wrapping the whole thing in a `loop`, and explicitly `return`ing when we encounter a well-formed pair. That way callers don't have to worry about handling cases the spec says are illegal.